apps/allocations/app/components/App/BudgetDetail.js

```diff
@@ -0,0 +1,242 @@
+
+const percentOf = (smaller, bigger) =>
+  `${BigNumber(100 * smaller / bigger).dp(1).toString()}%`
+
+const formatDate = date =>
+  new Intl.DateTimeFormat(undefined, {
+    dateStyle: 'long',
+    timeStyle: 'long',
+  }).format(date)
+
+function CurrencyValue({ amount, context, token }) {
+  const theme = useTheme()
+
+  const main = BigNumber(amount)
+    .div(10 ** token.decimals)
+    .dp(3)
+    .toNumber()
+    .toLocaleString()
+
+  return (
+    <>
+      <Text.Block>
+        <Text size="xlarge">
+          {main + ' '}
+        </Text>
+        <Text size="small">{token.symbol}</Text>
+      </Text.Block>
+      <Text.Block size="small" color={`${theme.surfaceContentSecondary}`}>
+        {context}
+      </Text.Block>
+    </>
+  )
+}
+
+CurrencyValue.propTypes = {
+  amount: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]).isRequired,
+  token: PropTypes.shape({
+    decimals: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]).isRequired,
+    symbol: PropTypes.string.isRequired,
+  }).isRequired,
+  context: PropTypes.string.isRequired,
+}
+
+function InfoBlock({ children, className, title }) {
+  const theme = useTheme()
+  return (
+    <div className={className}>
+      <div css={`
+          ${textStyle('label2')};
+          margin-bottom: ${GU}px;
+          color: ${theme.surfaceContentSecondary};
+        `}
+      >
+        {title}
+      </div>
+      {children}
+    </div>
+  )
+}
+
+InfoBlock.propTypes = {
+  children: PropTypes.node,
+  className: PropTypes.string,
+  title: PropTypes.string,
+}
+
+function usePeriod() {
+  const { appState } = useAragonApi()
+  if (!appState.period) return {}
+
+  let { startDate, endDate } = appState.period
+  const duration = endDate - startDate
+
+  // if user visits page from the timeframe of what should be the *next*
+  // period, but the period has not yet been updated on-chain, show them the
+  // period as it will be calculated for their date
+  while (endDate < new Date()) {
+    startDate = new Date(endDate.getTime() + 1000)
+    endDate = new Date(startDate.getTime() + duration)
+  }
+
+  return { startDate, endDate }
+}
+
+const Grid = styled.div`
+  display: grid;
+  grid-gap: ${2 * GU}px;
+  grid-template-areas:
+    "main"
+    "budget"
+    ${p => p.active && '"period"'}
+    "allocations";
+
+  @media (min-width: ${BREAKPOINTS.large}px) {
+    grid-template-columns: 3fr 1fr;
+    grid-template-rows: repeat(3, auto);
+    grid-template-areas:
+      "main budget"
+      "main period"
+      "allocations period";
+  }
+`
+
+export default function BudgetDetail() {
+  const { appState } = useAragonApi()
+  const { parsePath, patientlyRequestPath, requestPath } = usePathHelpers()
+  const { newAllocation } = usePanel()
+  const period = usePeriod()
+  const theme = useTheme()
+
+  const { id } = parsePath('/budgets/:id')
+
+  const budget = appState.budgets.find(b => b.id === id)
+  if (!budget) {
+    patientlyRequestPath('/')
+    return null
+  }
```

Comment:
This should be handled before loading the BudgetDetails page. Say a user is loading the link for the first time, they enter '/budgets/1' into the url, but `appState.budgets` hasn't synced yet. In this code, `if (!budget)` will return true, and the url will get reset to '/', so the user will be brought to the overview page, even if '/budgets/1' exists. If you don't call `requestPath('/')` then the url won't be reset and the page will load properly once `appState.budgets` has synced.

Reply:
The `patiently` part means that it waits one second for appState to get set up properly before redirecting. In my experience (and it sounds like in Kevin's), this one second delay is sufficient for getting the app state loaded correctly, even when loading this link for the first time. See definition of `patientlyRequestPath` for details on how it works.